File: TerraObserver/Scenes/Chunks/Views/HexGridChunk.cs
using System;
using System.Collections.Generic;
using Godot;
using TO.Domains.Types.Chunks;
using TO.Domains.Types.HexSpheres.Components.Chunks;

namespace TerraObserver.Scenes.Chunks.Views;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-06 16:31:36
[Tool]
public partial class HexGridChunk : Node3D, IHexGridChunk
{
    #region 事件

    public event Action<HexGridChunk>? Processed;

    #endregion

    #region Export 属性

    [Export] public HexMesh? Terrain { get; set; }
    public IHexMesh? GetTerrain() => Terrain;
    [Export] public HexMesh? Rivers { get; set; }
    public IHexMesh? GetRivers() => Rivers;
    [Export] public HexMesh? Roads { get; set; }
    public IHexMesh? GetRoads() => Roads;
    [Export] public HexMesh? Water { get; set; }
    public IHexMesh? GetWater() => Water;
    [Export] public HexMesh? WaterShore { get; set; }
    public IHexMesh? GetWaterShore() => WaterShore;
    [Export] public HexMesh? Estuary { get; set; }
    public IHexMesh? GetEstuary() => Estuary;
    [Export] public HexMesh? Walls { get; set; }
    public IHexMesh? GetWalls() => Walls;

    #endregion

    #region 普通属性

    public int Id { get; set; }
    public ChunkLodEnum Lod { get; set; } = ChunkLodEnum.JustHex;
    public HashSet<int>? EditingTileIds => null;

    #endregion

    #region 生命周期

    public override void _Process(double delta) => Processed?.Invoke(this);

    #endregion
}

File: TerraObserver/TO.Domains/Functions/Chunks/HexGridChunk.fs
namespace TO.Domains.Functions.Chunks

open Godot
open TO.Domains.Types.Chunks
open TO.Domains.Types.HexMeshes
open TO.Domains.Types.HexSpheres.Components

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-30 16:21:30
module HexGridChunkCommand =
    let clearOldData (env: 'E when 'E :> IFeatureCommand and 'E :> ITileQuery) (this: IHexGridChunk) =
        this.GetTerrain() |> HexMeshCommand.clear
        this.GetRivers() |> HexMeshCommand.clear
        this.GetRoads() |> HexMeshCommand.clear
        this.GetWater() |> HexMeshCommand.clear
        this.GetWaterShore() |> HexMeshCommand.clear
        this.GetEstuary() |> HexMeshCommand.clear
        this.GetWalls() |> HexMeshCommand.clear

        for tile in env.GetTilesByChunkId this.Id do
            env.HideFeatures tile.Id false
            env.DeleteFeatures tile.Id false

    let hideOutOfSight env (this: IHexGridChunk) =
        this.Hide()
        clearOldData env this
        this.Id <- 0 // 重置 id，归还给池子

    let applyNewData (this: IHexGridChunk) =
        this.GetTerrain() |> HexMeshCommand.apply
        this.GetRivers() |> HexMeshCommand.apply
        this.GetRoads() |> HexMeshCommand.apply
        this.GetWater() |> HexMeshCommand.apply
        this.GetWaterShore() |> HexMeshCommand.apply
        this.GetEstuary() |> HexMeshCommand.apply
        this.GetWalls() |> HexMeshCommand.apply

    let showMesh (meshes: Mesh array) (this: IHexGridChunk) =
        this.GetTerrain() |> HexMeshCommand.showMesh meshes[int MeshType.Terrain]
        this.GetWater() |> HexMeshCommand.showMesh meshes[int MeshType.Water]
        this.GetWaterShore() |> HexMeshCommand.showMesh meshes[int MeshType.WaterShore]
        this.GetEstuary() |> HexMeshCommand.showMesh meshes[int MeshType.Estuary]

    let getMeshes (this: IHexGridChunk) =
        [| this.GetTerrain().Mesh
           this.GetWater().Mesh
           this.GetWaterShore().Mesh
           this.GetEstuary().Mesh |]
